src/algebra/algebra/basic.lean

```diff
@@ -1053,244 +1056,229 @@ by rw [algebra_compatible_smul A r (a • m), smul_smul, algebra.commutes, mul_s
+  m ∈ V.restrict_scalars R ↔ m ∈ V :=
 iff.refl _
 
-@[simp]
-lemma submodule.restrict_scalars_bot :
-  submodule.restrict_scalars R (⊥ : submodule S E) = ⊥ :=
-rfl
+variables (R A M)
+
+lemma restrict_scalars_injective :
+  function.injective (restrict_scalars R : submodule A M → submodule R M) :=
+λ V₁ V₂ h, ext $ by convert set.ext_iff.1 (ext'_iff.1 h); refl
+
+@[simp] lemma restrict_scalars_inj {V₁ V₂ : submodule A M} :
+  restrict_scalars R V₁ = restrict_scalars R V₂ ↔ V₁ = V₂ :=
+⟨λ h, restrict_scalars_injective R _ _ h, congr_arg _⟩
 
 @[simp]
-lemma submodule.restrict_scalars_top :
-  submodule.restrict_scalars R (⊤ : submodule S E) = ⊤ :=
-rfl
+lemma restrict_scalars_bot : restrict_scalars R (⊥ : submodule A M) = ⊥ := rfl
 
-/-- The `R`-linear map induced by an `S`-linear map when `S` is an algebra over `R`. -/
-def linear_map.restrict_scalars (f : E →ₗ[S] F) :
-  (restrict_scalars R S E) →ₗ[R] (restrict_scalars R S F) :=
-{ to_fun := f.to_fun,
-  map_add' := λx y, f.map_add x y,
-  map_smul' := λc x, f.map_smul (algebra_map R S c) x }
+@[simp]
+lemma restrict_scalars_top : restrict_scalars R (⊤ : submodule A M) = ⊤ := rfl
 
-@[simp, norm_cast squash] lemma linear_map.coe_restrict_scalars_eq_coe (f : E →ₗ[S] F) :
-  (f.restrict_scalars R : E → F) = f := rfl
+end submodule
 
 @[simp]
-lemma restrict_scalars_ker (f : E →ₗ[S] F) :
+lemma linear_map.restrict_scalars_ker (f : M →ₗ[A] N) :
   (f.restrict_scalars R).ker = submodule.restrict_scalars R f.ker :=
 rfl
 
-/-- `A`-linearly coerce a `R`-linear map from `M` to `R` to a function, given an algebra `A` over
-a commutative semiring `R` and `M` a semimodule over `R`. -/
-def linear_map.lto_fun (R : Type u) (M : Type v) (A : Type w)
-  [comm_semiring R] [add_comm_monoid M] [semimodule R M] [comm_ring A] [algebra R A] :
-  (M →ₗ[R] A) →ₗ[A] (M → A) :=
-{ to_fun := linear_map.to_fun,
-  map_add' := λ f g, rfl,
-  map_smul' := λ c f, rfl }
-
 end semimodule
 
-section module
-
-instance (R : Type*) (S : Type*) (E : Type*) [I : add_comm_group E] :
-  add_comm_group (semimodule.restrict_scalars R S E) := I
-
-end module
-
 end restrict_scalars
 
+namespace linear_map
 section extend_scalars
 /-! When `V` is an `R`-module and `W` is an `S`-module, where `S` is an algebra over `R`, then
 the collection of `R`-linear maps from `V` to `W` admits an `S`-module structure, given by
-multiplication in the target -/
+multiplication in the target. -/
 
 variables (R : Type*) [comm_semiring R] (S : Type*) [semiring S] [algebra R S]
   (V : Type*) [add_comm_monoid V] [semimodule R V]
-  (W : Type*) [add_comm_monoid W] [semimodule S W]
+  (W : Type*) [add_comm_monoid W] [semimodule R W] [semimodule S W] [is_scalar_tower R S W]
 
 /-- The set of `R`-linear maps admits an `S`-action by left multiplication -/
-instance linear_map.has_scalar_extend_scalars :
-  has_scalar S (V →ₗ[R] (semimodule.restrict_scalars R S W)) :=
+instance has_scalar_extend_scalars :
```

Comment:
Same question as for `algebra_has_scalar`.

Reply:
This is a tricky situation.

Reply:
It is defeq to the old instance. I added an example that checks this.